mono/metadata/appdomain.c

```diff
@@ -820,6 +820,8 @@ mono_domain_has_type_resolve (MonoDomain *domain)
 	if (field == NULL) {
 		field = mono_class_get_field_from_name_full (mono_defaults.appdomain_class, "TypeResolve", NULL);
 		g_assert (field);
+	} else {
+		mono_memory_read_barrier ();
```

Comment:
Shouldn't the barriers be added to write not the read ?

Reply:
Both are needed I believe.
And I added both, in a fashion.
I added barrier in  `mono_class_get_field_from_name_full` and similar, about three of them.
Which is a little bit cheating.

The less cheating form would be like:
```
static type* static_foo;
type* foo = static_foo;
if (!foo) {
 foo = new_foo ();
 barrier; // or write barrier
 static_foo = foo; // maybe interlocked compare exchange
} else
  read_barrier
```

 1. this was getting tedious 
 2. diff was getting larger 
 3. I tried generalizing/abstracting it, which mostly is easy, but there is variation here and there, like asserts 
The lack of generalization/abstraction makes it easy to conform to every existing variation.

so I err'ed toward a smaller abtraction-less slightly less efficient change.

I'm definitely open to revisit that. With the ideas being, 1. the cache-less paths, do not need barriers. 2. the caching paths should be small and relatively repetition-free.

We first need to agree where the barriers should be.
Which I think this PR shows, except, again, it has a few extra, because it has write (or full) barriers in cache-free paths.

Reply:
Recall the general rule that barriers come in pairs. 
They are needed for reader and writer, else they are incomplete.

Reply:
Our current pattern used in the runtime code is:
if (!global) {
   local = init ();
  memory barrier
  global = local;
}

Reply:
Yes and that might not suffice.
Barriers are supposed to come in pairs.
For correctness, not style.

The exception might be “data dependency” that some/most processors recognize.

But recent experience with ARM does not meet that expectation and JIT now inserts more barriers than even I thought needed.

Reply:
Oh and maybe you are saying, I should make the extra changes then here, to take the barriers out of the overly common paths. I can do that.

Reply:
But the pattern should still say “else read_barrier”.

Reply:
Ok I started moving the write/full bariers out to the callers, and using the local/global split (or foo/static_foo, they are local either way).
There are around 100 edits to make this way so I"m not done yet.

Reply:
> The exception might be “data dependency” that some/most processors recognize.
> 
> But recent experience with ARM does not meet that expectation and JIT now inserts more barriers than even I thought needed.

To satisfy my curiosity, could you share an example of when your expectations around data dependencies/control dependencies from a read to a write weren't met? This is not a comment about this changeset, which I still need to read.

Reply:
The test case that was fixed by this, I think:
https://github.com/mono/mono/pull/17093

Reply:
@jaykrell @imhameed AFAIK, #17093 was about respecting the _CLR_ memory model (or more specifically... what CoreCLR actually does and what CoreFX assumes) on ARM, than making up for some data dependency issue in ARM hardware.  The issue is that CoreFX code assumes that some writes to static fields have release semantics.  So the PR is fixing up how mono behaves to match what CoreFX expects (because we imported a ton of CoreFX code, and because netcore Mono runs unaltered CoreFX code).  @BrzVlad did I get that right?

So separately from that, there's our C code.  Let's assume we're doing the totally wrong thing for Alpha (ie, we assume that data dependencies are respected and we don't use any sort of macros to add barriers where we rely on data dependencies).

So the other question is: do we need read barriers paired up with write barriers when we're doing our initialization pattern for statics.  I think the answer is "yes, technically, but in most places we don't use `MonoMethod*` or `MonoClass*` as a flag.

The issue is code like:

```
   static MonoMethod *static_method;
   MonoMethod * method = static_method;
   if (G_UNLIKELY (!method)) {
      method = init(klass); /* Suppose this also has a side effect of writing to klass->SomeMember */
      write_barrier ();
      static_method = method;
  } else {
     read_barrier ();
  }
  g_assert (method);
  use (klass->SomeMember, method);
```

without the read barrier, the read from `klass->SomeMemer` in the call to `use` could be reordered to the beginning of the method , so we could see method == NULL and klass->SomeMember with some junk value.

So yea, we need the read barrier in this initialization pattern.  I think most of the time we don't use the `static_method` as a flag to guarantee that other things got inited.  But I guess it's better to be safe than sorry.

Reply:
@lambdageek makes a point I had missed.
I was thinking, only “method” is read and written (indirect). Data dependency should handle that.

If as well “klass” is read/written (indirect) then yes, problem.
 
I wonder if dependencies can be induced, like say (klass - method + methor)->klassmember.

Reply:
@lambdageek makes a point I had missed.
I was thinking, only “method” is read and written (indirect). Data dependency should handle that.

If as well “klass” is read/written (indirect) then yes, problem.
 
I wonder if dependencies can be induced, like say (klass - method + methor)->klassmember.

Reply:
Artificial dependencies can be used at the ISA level; dependency tracking there is syntactic. Consider the pidgin assembly: `load dep, flag; xor dep, dep; load val, (guarded + dep)`; the load from `guarded` would be ordered after the load from `flag`. You'd need something like http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0750r1.html to encode this in a higher level language with optimizations.

Reply:
Yeah the ARM manual mentions such things.
Maybe we can encode that in C?
```
void mono_address_dependency (void * volatile *a, void* volatile dep)
{
  *(char**)a += (size_t)dep - (size_t)dep;
}
```